### ISSUE TITLE
Lagging brush strokes in MyPaint

### ISSUE THREAD

**Issue by AtsusaKaneytza** [mypaint/mypaint#878]
#### Description of the problem

There's some very annoying lag in brush strokes in MyPaint where it hasn't existed before. It seems to coincide with the GTK fix for the pen pressure issues.

#### Basic system details

MyPaint version: 1.3.0-alpha+git.050d44f4, 64-bit
Operating system: Windows 10, 64-bit
Microsoft Surface 3, 2GB Ram

#### Steps to reproduce

Not really sure how to explain how to reproduce this error: this brush lag also happens in any 32-bit version I test that meets the aforementioned criteria as well.

1. Start MyPaint
2. Attempt to draw any quick strokes
3. proceed to notice the agonizing lag

#### Backtraces or error messages

Not sure what error messages or input values I should post here, but I would appreciate feedback on what to post to help figure out the error.

**Comment by geonom** [comment]
Having a similar issue with the lags:

On my Surface Pro 4 (i7, 8 GB) in Windows 10, mypaint brush strokes behave differently according to myspace **window size** (smaller window size means better performance, less lag), and **brush complexity(?)** - meaning the deevad/2B pencil is much faster (even with window fullsize nearly no lag) than the deevad/basic brush, which lags a lot when the mypaint window is fullsize but nearly not at all when the window is sized about 1/4 of the screen .

This is what the debug screen shows (mypaint-git-w64-1.3.0-alpha-master.178-installer):

`Running python2.exe...
INFO: mypaint: Installation layout: conventional POSIX-like structure with prefix u'C:/Program Files/mypaint-git-w64/mingw64'
INFO: lib.i18n: Windows: LANG='en_GB'
INFO: lib.i18n: Windows: LANGUAGE='en_GB:de_DE'
INFO: lib.i18n: Value of LANGUAGE after cleanup: 'en_GB:C:de_DE'
WARNING: mypaint: No bindtextdomain builtins found in module 'locale'.
INFO: mypaint: Trying platform-specific fallback hacks to find bindtextdomain funcs.
INFO: mypaint: Windows: found working bindtextdomain funcs in 'libintl-8.dll' (ctypes)
C:/Program Files/mypaint-git-w64/mingw64/lib/mypaint/gui/overlays.py:19: PyGIWarning: PangoCairo was imported without specifying a version first. Use gi.require_version('PangoCairo', '1.0') before import to ensure that the right version gets loaded.
  from gi.repository import PangoCairo
WARNING: gui.workspace: win32: floating windows are not supported (bug #679), not enabling subwindow creation by snapping out.
WARNING: gui.workspace: win32: floating windows are not supported (bug #679), not enabling subwindow creation by snapping out.
INFO: gui.device: New device 'System Aggregated Pointer' (GDK_SOURCE_MOUSE, axes:2, class=GdkDeviceWin32, vendor=None, product=None)
INFO: gui.document: Initialized background from u'C:/Program Files/mypaint-git-w64/mingw64/share/mypaint/backgrounds/10_soft_yellow.png'
WARNING: gui.keyboard: Ignoring keybinding for '<Actions>/BrushModifierActions/BlendModeMenu'

(python2.exe:6800): Gtk-WARNING **: Theme parsing error: <data>:5:50: The style property GtkMenuBar:internal-padding is deprecated and shouldn't be used anymore. It will be removed in a future version
WARNING: gui.workspace: win32: floating windows are not supported (bug #679), not enabling subwindow creation by snapping out.
INFO: lib.document: load_ora: u'C:/Users/rosa arvensis/AppData/Local/mypaint/scratchpads/autosave.ora'
INFO: lib.document: 0.315s load_ora total
INFO: gui.filehandling: Loaded scratchpad from u'C:/Users/rosa arvensis/AppData/Local/mypaint/scratchpads/autosave.ora'
WARNING: lib.tiledsurface: empty surface
INFO: gui.device: New device 'Microsoft device Stylus' (GDK_SOURCE_PEN, axes:3, class=GdkDeviceWintab, vendor=None, product=None)`

**Comment by AtsusaKaneytza** [comment]
Still waiting on feedback on what tablet input info I should share and how, regarding this problem. I don't want to clutter the page with irrelevant info.

@geonom how did you share your input values, and how do I do the same?

**Comment by kelvinwop** [comment]
I've found that the first inch or so of the stroke never lags while after about three inches, it will almost always (90% ish) lag slightly before catching up.
smoothing is set to 0.

**Comment by AtsusaKaneytza** [comment]
The latest stable version's (1.2.1) 32 and 64-bit versions also have this issue on my main work laptop (it has 8 total GB of RAM, with 4GB being VRam) Laptop runs windows 10 64-bit
I have an XP-Pen artist 13 laptop, and the issue persists, regardless if MyPaint is the only software open or not. I've also tried running MyPaint with my dedicated graphics, to no effect.

**Comment by kelvinwop** [comment]
Check your device usage like in #901, making sure multiple things weren’t inputting at the same time got rid of tons of lag

**Comment by Trokinos** [comment]
It is a very strange problem as i observe it on no other of my painting/graphic applications (krita 4.x ,  gimp 2.8.x , firealpaca, blender paint mode).
On my window 10 system and wacom intuos i noticed it does not happen in Mypaint 1.0 in which every brush work perfectly with every brush strokes, but in every further version i tried the difference in brush lagging/delay, the feel of them is more or less noticable.

And it seems the problem is impacted by the device stuff mentionned by kelvinwop, as on a test i noticed my wacom wasn't even listed (only the system aggregated thingy was in there) and the lag was too much to be even usable.

On the current version i'm testing, the latest at the time from the appveyor build (mypaint-git-w64-2 0 0-alpha-master 482-installer) , both the wacom and the system aggregated are listed (was it something that was fixed or just pure luck), somehow the brushes seemed to have very few lag delay (still not as hyper responsive as they were in MyPaint 1.0 unfortunately)  i wondered what happened if i turned the system aggregated to "Ignore" (as it was the mouse and while painting i didn't really needed it) and while i was expecting improvement instead immediately the brush lag became unbearable.

Turned it back to "All Task" and it went back to normal (slightly laggy in comparison to 1.0), it does not make much sense to me.

I also noticed in every brush tool the "Pigment" slider is full on for all brushes i tried, for the most simpler brush it does not show, but for more complex/large brush it's unworkable laggy until i set this Pigment to 0 in the brush tool in which situation the brush performance is largely improved.

It seems this problem affect the windows build of Mypaint since a very long time as from my google search all those similar brush laggy problem are always coming from window mypaint users, on linux it seems to work without problem apparently.
But at least a way to turn off Pigment globally would be usefull (as it's a pain to go through each brush individually to do that) at least until this laggy brush on window problem can find a solution

**Comment by 1vanK** [comment]
https://youtu.be/if0OE-lcVK4

Blue lines drawed with tablet XP-PEN Star G640
Red lines drawed with mouse (line end is drawn after releasing LMB)

MyPaint version: mypaint-git-w64-2.0.0-alpha-master.502-standalone.zip
OS: Windows 10

**Comment by jplloyd** [comment]
Thanks for the demo Ivan, those strokes certainly seem slower than they should be (unless the computer it is running on is very slow).

The "delay" on the strokes painted with the mouse is not really lag, but (I'm fairly certain) a consequence of the fact that the brush you used is using per-dab tracking with a high value, and that your mouse does not send enough position events for the stroke to keep going.

**Comment by 1vanK** [comment]
Thanks for the clarification! Dab settings helped me

**Comment by AesaraB** [comment]
Potentially a duplicate of #803, but that one's specific to Windows.

**Comment by BrunoDSL** [comment]
> @parhizkari If you have upgraded to MyPaint >= 2.0.0, I would suggest going to the compatibility settings and setting the default mode to `1.x`. The new default blending mode in 2.0.0 is a lot heavier than the default in 1.2.1 and older. Even so, the hardware may be a bit weak to run newer versions of MyPaint, though the next release will include a number of performance improvements.
 
I just tested here with a Ryzen 1600AF, 8GB of RAM and a 4GB RX550 and there's no difference in the delay under Windows. Haven't tested under Linux. Can't test with current development versions, as the last successful nightly build has been deleted.
